add github actions with powershell scripts (#5724)
umbraco/UmbracoDocs: umbraco-cloud/set-up/project-settings/umbraco-cicd/samplecicdpipeline/azure-devops-pwsh.md

---
description: >-
  This section provides a step-by-step guide to setting up a CI/CD pipeline in
  Azure DevOps using the provided sample Powershell scripts.
---

# Azure DevOps with Powershell scripts

> Before setting up the pipeline in Azure Devops, make sure that the steps in the [Configuring a CI/CD pipeline](./) are done.

Reviewer: DevOps

You will need to define your pipeline in YAML, and find a way to interact with the Umbraco Cloud Api.

> {% hint style="info" %}

Reviewer: this part isn't rendered correctly in the markdown preview

Author: is it the {% hint ....%} part?I guess this is GitBook magic :) It should look correct once merged into docs

> We have created a sample pipeline for Azure Devops.  It includes YAML-files and custom PowerShell scripts to interact with the Umbraco Cloud API.

Reviewer: DevOps

> You can get the samples for Azure Devops and GitHub from this [repository](https://GitHub.com/umbraco/Umbraco.Cloud.CICDFlow.Samples).

Reviewer: DevOps

> For Azure Devops you will need the following files:

Reviewer: DevOps

- From `powershell` folder
  - `Get-LatestDeployment.ps1`
  - `Get-ChangesById.ps1`
  - `New-Deployment.ps1`
  - `Add-DeploymentPackage.ps1`
  - `Start-Deployment.ps1`
  - `Test-DeploymentStatus.ps1`
- From the `powershell/azuredevops` folder
  - `azure-release-pipeline.yml`
  - `cloud-sync.yml` 
  - `cloud-deployment.yml`

> - From the root folder

Reviewer: I think this should be the first item in the list due to the folder hierarchy

Author: Good point - I also did it in the github example

>   - `cloud-deployment.yml`

Author: yes this is much better with your suggestion! :)

{% endhint %}

> ## Import Cloud project repository to Azure Devops

Reviewer: DevOps

> Go to your repositories in Azure Devops and click on "Create a repository".

Reviewer: DevOps

- Create a new empty repository (don't add a README and don't add a .gitignore), and note down the clone URL.
- Go to the Umbraco Cloud Portal and clone your cloud project down locally. [This article](../../../working-locally.md#cloning-an-umbraco-cloud-project) describes how you can find the clone URL.

> - Now working locally remove the Git Remote called `origin``, which currently points to Umbraco Cloud

Reviewer: I think there is a typo here with `origin`` the backticks don't match

 ```sh 
 git remote remove origin
 ```

- Optionally rename branch `master` to `main`

 ```sh 
 # optional step
 git branch -m  main
 git symbolic-ref HEAD refs/heads/main
 ```

- Add a new remote called origin and pointing to the Azure Devops clone URL and push

 ```sh 
 git remote add origin https://{your-organization}@dev.azure.com/{your-organization}/{azure-project-scope}/_git/{your-repository}
 git push -u origin --all
 ```

Now we can move on to setting up a pipeline.

> ## Set up the Azure Devops pipeline files

Reviewer: DevOps

While working with the project on you local machine, do the following to prepare the pipeline.
- Copy the `cloud.zipignore` file to the root of your repository
- Make a copy of the `.gitignore` and call it `cloud.gitignore`

>   - Both files should be in the root of your repository

Author: note for Jesper so he remembers:
add a line about gitignore the git-patch.diff file non cloud .gitignore

- Also in the root, create a folder called `devops`
- Copy the 3 YAML files into the `devops` folder
- Inside `devops` create an additional folder called `powershell`
- Copy the Powershell scripts to the `powershell` folder
- Commit the all changes, and push to Azure Devops

> ## Configure Azure Devops

Reviewer: DevOps

The pipeline needs to know which Umbraco Cloud project to deploy to. In order to do this you will need the `Project ID` and the `API Key`. [This article](README.md#obtaining-the-project-id-and-api-key) describes how to get those values.

- Now go to the repository in Azure and click on "Set up build".

<figure><img src="../../../../.gitbook/assets/azuresetupbuild.png" alt=""><figcaption><p>Azure Devops Repository</p></figcaption></figure>

- On the next screen click on "Existing Azure Pipelines YAML file"

<figure><img src="../../../images/Pipeline3.png" alt=""><figcaption><p>Configure pipeline with existing YAML file</p></figcaption></figure>

- Select `main` (or `master` if you did not change the branch name) in Branch

> - Select `/devops/azure-release-pipelin.yaml` in Path and continue

Reviewer: pipelin typo 😄

<figure><img src="../../../images/Pipeline4.png" alt=""><figcaption><p>Select Branch and Path</p></figcaption></figure>

- Now you are on the "Review your pipeline YAML" screen
  - Replace the `##Your project Id here##` with the Project Id you got from Umbraco Cloud Portal
  - Click on "Variables"
  - Add the variable `umbracoCloudApiKey` with the value of the API Key you got from Umbraco Cloud Portal

> {% hint style="info" %}

Reviewer: this isn't rendered correctly

> If you want to use other names for the variable, you need to rename the affected variables in `azure-release-pipelin.yaml`.

Reviewer: pipelin typo 😄

{% endhint %}

When you click on "Run" your first deployment will be triggered. Which means that Azure Devops is set up with all the needed information to be able to deploy your Cloud project back to Umbraco Cloud.

## High level overview of the pipeline components

> The scripts are provided as a starting point. It is recommended that you familiarize yourself with the scripts and with documentation related to how to use Azure Devops.

Reviewer: DevOps

> The scripts demonstrates two thing:

Reviewer: the following: 
maybe sounds better

Author: I changed it to "The mentioned scripts are provided...." which I find even better :D what do you think? :)

>  - How to sync your Azure Devops repository with the left-most project environment in Umbraco Cloud

Reviewer: DevOps

 - How to deploy changes to the left-most project environment in Umbraco Cloud 

### Main

> The `azure-release-pipelin.yaml` is the main pipeline, and is the one that will be triggered on a push to `main` branch.

Reviewer: pipelin 😄

> You can configure different trigger behavior in this file.

Reviewer: a different trigger behavior

You can add your Build and Test stage between the `cloudSyncStage` and `cloudDeploymentStage` stages. 
Keep in mind that you do not need to retain the dotnet build artifact for upload later. The `cloudDeploymentStage` job will take care of packaging all your source code and upload to Umbraco Cloud. 

### Cloud-sync

> The `cloud-sync.yml` shows how you can sync your Azure Devops repository with the left-most environment of your Cloud project.

Reviewer: DevOps

> In this sample, it accepts any change from the api and applies and commits it back to the branch which triggered the pipeline. But the commit does not trigger the pipeline again.

Reviewer: you can say "However, " instead of But

Author: very nice

If you don't want the pipeline to commit back to the triggering branch, this is where you need to change the pipeline. 

> ### Cloud-deployment

Reviewer: Cloud deployment

> The `cloud.deployment.yml` show how you can deploy your repository to the left-most environment of your Cloud project.

Reviewer: shows*

> The `cloud.deployment.yml` show how you can deploy your repository to the left-most environment of your Cloud project.

Reviewer: cloud-deployment

> The sample shows how to prepare for deployment, request the deployment and await for cloud to finish.

Reviewer: wait for Cloud 
or
await Cloud

There are a couple of things here to be aware of:
- We are overwriting the `.gitignore` with the `cloud.gitignore`.
  This is a way to accommodate your gitignore-needs when working locally. For instance you might want to ignore frontend builds, but you want them build and published to cloud.  
- We have a special `cloud.zipignore`.

>   This is a convenient way to tell the pipeline which files **not** to include when creating the zip package to send to cloud.

Reviewer: cloud or Cloud?

> If you have frontend assets that needs to be build (using tools like npm/yarn or others), you should add the needed steps before `Zip Source Code`. This is to ensure that the fresh frontend assets will be part of the package to be sent to cloud. 

Reviewer: needs to be built

Reviewer: cloud or Cloud?

---
description: >-
  This section provides a step-by-step guide to setting up a CI/CD pipeline in
  Azure DevOps using the provided sample Powershell scripts.
---

# Azure DevOps using Powershell scripts

Before setting up the pipeline in Azure DevOps, make sure that the following steps from the [Configuring a CI/CD pipeline](./) are done:
- Use an existing Cloud project or create a new one
- Activate CI/CD Flow

Next you will need to define your pipeline in YAML, and use it to interact with the Umbraco Cloud API.

{% hint style="info" %}
The Umbraco CI/CD Team have created a sample pipeline for Azure DevOps. 

The Scripts are provided as is. This means that the scripts will do the bare minimum for a pipeline that is utilizing the CI/CD flow. Adapt and integrate them into your own pipelines to gain the ability to do deployments to your Umbraco Cloud projects.

The sample includes YAML-files and custom Powershell scripts to interact with the Umbraco Cloud API.

> You can get the samples for Both `Azure DevOps` and `GitHub` from [here](https://github.com/umbraco/Umbraco.Cloud.CICDFlow.Samples).

Reviewer 2: ```suggestion
You can get the samples for both `Azure DevOps` and `GitHub Actions` from the [Github repository](https://github.com/umbraco/Umbraco.Cloud.CICDFlow.Samples).
```

  - Add the variable `umbracoCloudApiKey` with the value of the API Key you got from Umbraco Cloud Portal

{% hint style="info" %}

It is recommended to handle the `API Key` as a secret. This can be done by ticking the "Keep this value secret" checkbox.  

<figure><img src="../../../../.gitbook/assets/azdevops-secret-variable.png" alt=""><figcaption><p>Make the variable secret in Azure DevOps</p></figcaption></figure>

{% endhint %}

{% hint style="info" %}
You can customize the names for the variables as you like, however you then need to rename the affected variables in `azure-release-pipeline.yaml`.

{% endhint %}

When you click on "Save and Run" your first deployment will be triggered. Which means that Azure DevOps is set up with all the needed information to be able to deploy your Cloud project back to Umbraco Cloud.

### Optional: Run a deployment to confirm

With everything set up, you may want to confirm that Umbraco Cloud reflect the changes you are sending via your pipeline.

While working on you project locally, ad a new Document type.
- Commit the change to `main` and push to your repository.
- The pipeline starts to run

> - Once the pipeline is done log into Backoffice on your left-most environment in cloud

Reviewer 2: ```suggestion
- Once the pipeline is done log into Backoffice on your left-most environment in Umbraco Cloud
```

You can configure a different trigger behavior in this file.

You can add your Build and Test stage between the `cloudSyncStage` and `cloudDeploymentStage` stages. 
Keep in mind that you do not need to retain the dotnet build artifact for upload later. The `cloudDeploymentStage` job will take care of packaging all your source code and upload to Umbraco Cloud. 

### Cloud-sync

The `cloud-sync.yml` shows how you can sync your Azure DevOps repository with the left-most environment of your Cloud project.
In this sample, it accepts any change from the API and applies and commits it back to the branch which triggered the pipeline. However the commit does not trigger the pipeline again.

If you don't want the pipeline to commit back to the triggering branch, this is where you need to change the pipeline. 

### Cloud-deployment

The `cloud-deployment.yml` shows how you can deploy your repository to the left-most environment of your Cloud project.
The sample shows how to prepare for deployment, request the deployment and wait for cloud to finish.

There are a couple of things here to be aware of:

> - We are overwriting the `.gitignore` with the `cloud.gitignore`.

Reviewer 2: ```suggestion
- We are overwriting the `.gitignore` file with `cloud.gitignore`.
```

  This is a way to accommodate your gitignore-needs when working locally. For instance you might want to ignore frontend builds, but you want them build and published to cloud.  

> - We have a special `cloud.zipignore`.

Reviewer 2: ```suggestion
- We have a special `cloud.zipignore` file.
```

  This is a convenient way to tell the pipeline which files **not** to include when creating the zip package to send to cloud.

> If you have frontend assets that needs to be built (using tools like npm/yarn or others), you should add the needed steps before `Zip Source Code`. This is to ensure that the fresh frontend assets will be part of the package to be sent to Umbraco Cloud. 

Reviewer 2: ```suggestion
{% hint style="info" %}

If you have frontend assets that needs to be built (using tools like npm/yarn or others), you should add the needed steps before `Zip Source Code`. This is to ensure that the fresh frontend assets will be part of the package to be sent to Umbraco Cloud. 

{% endhint %}
```

{% endhint %}

## Import Cloud project repository to Azure DevOps
Go to your repositories in Azure DevOps and click on "Create a repository".

- Create a new empty repository (don't add a README and don't add a .gitignore), and note down the clone URL.
- Go to the Umbraco Cloud Portal and clone your cloud project down locally. [This article](../../../working-locally.md#cloning-an-umbraco-cloud-project) describes how you can find the clone URL.
- Now working locally remove the Git Remote called `origin`, which currently points to Umbraco Cloud

 ```sh 
 git remote remove origin
 ```

- Optionally rename branch `master` to `main`

 ```sh 
 # optional step
 git branch -m  main
 git symbolic-ref HEAD refs/heads/main
 ```

- Add a new remote called origin and pointing to the Azure DevOps clone URL and push

 ```sh 
 git remote add origin https://{your-organization}@dev.azure.com/{your-organization}/{azure-project-scope}/_git/{your-repository}
 git push -u origin --all
 ```

Now we can move on to setting up a pipeline.

## Set up the Azure DevOps pipeline files

> While working with the project on you local machine, follow these steps to prepare the pipeline, using the [samples from the repository](https://github.com/umbraco/Umbraco.Cloud.CICDFlow.Samples).

Reviewer 2: ```suggestion
While working with the project on your local machine, follow these steps to prepare the pipeline, using the [samples from the repository](https://github.com/umbraco/Umbraco.Cloud.CICDFlow.Samples).
```
